src/cmp.rs

```diff
@@ -0,0 +1,1188 @@
+    let Ok(dev_null) = fs::metadata("/dev/null") else {
+        return false;
+    };
+
+    let stdout_fd = io::stdout().lock().as_raw_fd();
+
+    // SAFETY: we have exclusive access to stdout right now.
+    let stdout_file = unsafe { fs::File::from_raw_fd(stdout_fd) };
+    let Ok(stdout) = stdout_file.metadata() else {
+        return false;
+    };
+
+    let is_dev_null = stdout.dev() == dev_null.dev() && stdout.ino() == dev_null.ino();
+
+    // Don't let File close the fd. It's unfortunate that File doesn't have a leak_fd().
+    std::mem::forget(stdout_file);
+
+    is_dev_null
+}
+
+pub fn parse_params<I: Iterator<Item = OsString>>(mut opts: Peekable<I>) -> Result<Params, String> {
+    let Some(executable) = opts.next() else {
+        return Err("Usage: <exe> <from> <to>".to_string());
+    };
+    let executable_str = executable.to_string_lossy().to_string();
+
+    let parse_skip = |param: &str, skip_desc: &str| -> Result<usize, String> {
+        let suffix_start = param
+            .find(|b: char| !b.is_ascii_digit())
+            .unwrap_or(param.len());
+        let mut num = match param[..suffix_start].parse::<usize>() {
+            Ok(num) => num,
+            Err(e) if *e.kind() == std::num::IntErrorKind::PosOverflow => usize::MAX,
+            Err(_) => {
+                return Err(format!(
+                    "{}: invalid --ignore-initial value '{}'",
+                    executable_str, skip_desc
+                ))
+            }
+        };
+
+        if suffix_start != param.len() {
+            // Note that GNU cmp advertises supporting up to Y, but fails if you try
+            // to actually use anything beyond E.
+            let multiplier: usize = match &param[suffix_start..] {
+                "kB" => 1_000,
+                "K" => 1_024,
+                "MB" => 1_000_000,
+                "M" => 1_048_576,
+                "GB" => 1_000_000_000,
+                "G" => 1_073_741_824,
+                "TB" => 1_000_000_000_000,
+                "T" => 1_099_511_627_776,
+                "PB" => 1_000_000_000_000_000,
+                "P" => 1_125_899_906_842_624,
+                "EB" => 1_000_000_000_000_000_000,
+                "E" => 1_152_921_504_606_846_976,
+                "ZB" => usize::MAX, // 1_000_000_000_000_000_000_000,
+                "Z" => usize::MAX,  // 1_180_591_620_717_411_303_424,
+                "YB" => usize::MAX, // 1_000_000_000_000_000_000_000_000,
+                "Y" => usize::MAX,  // 1_208_925_819_614_629_174_706_176,
+                _ => {
+                    return Err(format!(
+                        "{}: invalid --ignore-initial value '{}'",
+                        executable_str, skip_desc
+                    ));
+                }
+            };
+
+            let (n, overflows) = num.overflowing_mul(multiplier);
+            if overflows {
+                num = usize::MAX;
+            } else {
+                num = n;
+            }
```

Comment:
With pattern matching this can be compacted to:
```rust
num = match num.overflowing_mul(multiplier) {
    (n, false) => n,
    _ => usize::MAX,
}
```